src/tools/apicall/apicall.ts

```diff
@@ -0,0 +1,131 @@
+ */
+
+import { join } from 'path';
+
+import { BaseToolOptions, StringToolOutput, Tool, ToolError } from '@/tools/base.js';
+import { SchemaObject } from 'ajv';
+import { parse } from 'yaml';
+import { isEmpty } from 'remeda';
+import axios from 'axios';
+import { HttpProxyAgent } from 'http-proxy-agent';
+import { HttpsProxyAgent } from 'https-proxy-agent';
+
+
+export interface ApiCallToolOptions extends BaseToolOptions {
+  name: string;
+  description?: string;
+  openApiSchema?: any;
+  apiKey?: string;
+  http_proxy_url?: string;
+}
+
+export class ApiCallTool extends Tool<StringToolOutput, ApiCallToolOptions> {
+  name: string;
+  description: string;
+  openApiSchema: any;
+  apiKey?: string;
+  http_proxy_url?: string;
+
+  inputSchema() {
+    return {
+      type: 'object',
+      required: ['path', 'method'],
+      oneOf: Object.entries(this.openApiSchema.paths).flatMap(([path, pathSpec]: [string, any]) =>
+        Object.entries(pathSpec).map(([method, methodSpec]: [string, any]) => ({
+          additionalProperties: false,
+          properties: {
+            path: {
+              const: path,
+              description:
+                'Do not replace variables in path, instead of, put them to the parameters object.'
+            },
+            method: { const: method, description: methodSpec.summary || methodSpec.description },
+            ...(methodSpec.requestBody?.content?.['application/json']?.schema
+              ? {
+                  body: methodSpec.requestBody?.content?.['application/json']?.schema
+                }
+              : {}),
+            ...(methodSpec.parameters
+              ? {
+                  parameters: {
+                    type: 'object',
+                    additionalProperties: false,
+                    required: methodSpec.parameters
+                      .filter((p: any) => p.required === true)
+                      .map((p: any) => p.name),
+                    properties: methodSpec.parameters.reduce(
+                      (acc: any, p: any) => ({
+                        ...acc,
+                        [p.name]: { ...p.schema, description: p.name }
+                      }),
+                      {}
+                    )
+                  }
+                }
+              : {})
+          }
+        }))
+      )
+    } as const satisfies SchemaObject;
+  }
+
+  public constructor({ name, description, openApiSchema, apiKey, http_proxy_url, ...rest }: ApiCallToolOptions) {
+    super({ name, description, ...rest });
+    this.name = name;
+    this.apiKey = apiKey;
+    this.http_proxy_url = http_proxy_url;
+    this.description = description ?? 'Use input schema to infer description';
+    this.openApiSchema = parse(openApiSchema);
+    // TODO: #105 review error codes. Were using APIErrorCode, now ToolError
+    if (!this.openApiSchema?.paths) {
+      throw new ToolError("Server is not specified!")
+    }
+  }
+
+  protected async _run(input: any) {
```

Comment:
I would try to use `protected async _run(input: SchemaObject) {` instead of the `any` type.